### ISSUE TITLE
Unable to re-produce SGD numbers

### ISSUE THREAD

**Issue by amitchandak** [timgaripov/swa#14]
Hi,
Thanks for SWA, neat idea. I just checked-out the code and wanted to replicate the SWA experiment as mentioned here . In Table1 of the paper you have mentioned SGD numbers are much higher than what I get when I run the code. Infact the SWA number I get is better. I also tried another minor change doing "bn_update"  operation even for SGD before evaluation and it improves the performance, not sure why ? I am attaching files of training VGG16BN model on CIFAR-10 dataset. Will be thankful for your insights.
[VGG16BN Model]
(https://github.com/timgaripov/swa/files/4008439/cifar10_VGG16BNModel_swaLogs.txt)

[VGG16BN Model with BN update for SGD](https://github.com/timgaripov/swa/files/4008443/cifar10_VGG16BNModel_sgdBN_swaLogs.txt)

**What's the best way to get similar or exact numbers as mentioned in paper**. I used the same command as mentioned in Git repo:
python3 train.py --dir=/home/swa/swa_cifar10_VGG16BNModel/ --dataset=CIFAR10 --data_path=/home/swa_gaussian/cifar10_data/ --model=VGG16BN --epochs=300 --lr_init=0.05 --wd=5e-4 --swa --swa_start=161 --swa_lr=0.01 --save_freq=50 --eval_freq=10 > cifar10_VGG16BNModel_swaLogs

Thank you once again.

**Comment by izmailovpavel** [comment]
Hi @amitchandak, thank you for your interest in our work :)

I don't think we reported results for VG16BN in the repo or in the paper. The numbers we reported are for VGG16, which doesn't have BatchNorm. It is expected that VGG16BN would perform **better**. 

As for getting good SGD results, you shouldn't use `--swa` and related flags. Try
```
python3 train.py --dir=<DIR> --dataset=CIFAR100 --data_path=<PATH> --model=VGG16BN --epochs=300 --lr_init=0.05 --wd=5e-4 # SGD
```

**Comment by amitchandak** [comment]
Thank you for your reply. Just to clarify, what do the numbers corresponding to columns  "te_loss  and   te_acc" mean then ? I thought these are the performance numbers for SGD model.

**Comment by izmailovpavel** [comment]
Those are the loss and accuracy of the SGD model, but when using the SWA learning rate schedule. The SWA learning rate schedule keeps the learning rate high until the very end of training which is good for SWA but suboptimal for SGD.

**Comment by amitchandak** [comment]
Thank you. Missed that point. Appreciate your help.

**Comment by amitchandak** [comment]
@izmailovpavel  I tried what you have suggested. But still when I run the experiment, the numbers don't match. 
Output of this command: 
python3 train.py --dir=/home/amit--dataset=CIFAR100 --data_path=/datapath/ --model=VGG16 --epochs=300 --lr_init=0.05 --wd=5e-4 --swa --swa_start=161 --swa_lr=0.01  > cifar100_VGG16Model_swaLogs
Last line of the file: " 300    0.0100     0.2326   93.6700     1.6885   66.6000         1.2728       74.1400   15.1881" . The accuracy of SWA model is **74.14** while in paper this number is **74.27** (Table 1 of SWA paper, UAI-2018) [cifar100_VGG16Model_swaLogs.txt](https://github.com/timgaripov/swa/files/4012611/cifar100_VGG16Model_swaLogs.txt)

Similarly for CIFAR-10, VGG16
python3 train.py --dir=/home/amit --dataset=CIFAR10 --data_path=/datapath/ --model=VGG16 --epochs=300 --lr_init=0.05 --wd=5e-4 --swa --swa_start=161 --swa_lr=0.01  > cifar10_VGG16Model_swaLogs
Last line of the file: "300    0.0100   0.0481   98.4840     0.3811   90.1400    0.2567   93.3700   14.8065" . The accuracy of SWA model is **93.37** while in paper this number is **93.64** (Table 1 of SWA paper, UAI-2018) .  [cifar10_VGG16Model_swaLogs.txt](https://github.com/timgaripov/swa/files/4012616/cifar10_VGG16Model_swaLogs.txt)

**How should I get same numbers as mentioned in paper  ?** Please find the log files attached for your reference.  

Thank you for your help

**Comment by izmailovpavel** [comment]
Hi @amitchandak. The difference between 74.14 and 74.27 is very minor. The exact numbers you get are different every time you re-train the model. Our results were averaged over 3 runs, and the variance is also reported. Your result is within one standard deviation.

**Comment by amitchandak** [comment]
Thank you for the clarification
